### PR TITLE
[NGS-833]  Add Multi Bidfloor Support For DSPs

### PR DESCRIPTION
**Pull request by mannychu**
## Context

We want to add support for multiple bid floors. Multi Bid Floor was previously handled in TPE Prebid, but we want to move this logic into MBS/TJX. 

Updated structs to support new bid floor fields 
Added support for multi bid floor traffic features
Updated tests to include new bid floor field

TPE Specific Changes:
Removed Pubmatic Multi Bid Floor handling and migrated into MBS/TJX
Overwrite the main imp bidfloor if multi bidfloor is present in the DSP imp ext

 ## How To Test

`make build` and `make test`
./validate.sh
End to end testing

 ## How Has This Been Tested?

`make build` and `make test`
./validate.sh
End to end testing

## Related PRs

Go Prebid: https://github.com/Tapjoy/go-prebid/pull/91
MBS: https://github.com/Tapjoy/mediation_bid_service/pull/1217
AS: https://github.com/Tapjoy/ad_service/pull/1030
TPEPS: https://github.com/Tapjoy/tpe_prebid_service/pull/122

 ## Jira Link

Jira: https://tapjoy.atlassian.net/browse/NGS-833

### CODE REVIEW

File: adapters/tapjoy/tapjoy.go [hunk 1]
```diff
@@ -185,6 +185,11 @@ func (a *adapter) MakeRequests(request *openrtb.BidRequest, _ *adapters.ExtraReq
 			thisImp.Banner = nil
 		}
 
+		// Overwrite BidFloor if present
```
Comment (umutdemirel) [inline adapters/tapjoy/tapjoy.go:188]: Logic I mentioned is above:
```
// overwrite bid floor with mediator given bid floor
thisImp.BidFloor = tapjoyExt.Imp.BidFloor

```

File: adapters/unicorn/unicorntest/exemplary/banner-app-with-ip.json [hunk 2]
```diff
@@ -149,7 +149,8 @@
                   "accountId": 199578,
                   "publisherId": 123456,
                   "mediaId": "test_media",
-                  "placementId": "test_placement"
+                  "placementId": "test_placement",
+                  "bid_floor": null
```
Comment (umutdemirel) [inline adapters/unicorn/unicorntest/exemplary/banner-app-with-ip.json:153]: Again, you can use omitempty for bid_floor to avoid updating here. And you should add one test for non-null bid_floor value if you can.

File: openrtb_ext/imp_tapjoy.go [hunk 3]
```diff
@@ -12,10 +12,11 @@ type ExtImpTapjoy struct {
...
 
-	Region         string `json:"region"`
-	Reward         int    `json:"reward"`
-	SKADNSupported bool   `json:"skadn_supported"`
-	MRAIDSupported bool   `json:"mraid_supported"`
+	Region         string   `json:"region"`
+	Reward         int      `json:"reward"`
+	SKADNSupported bool     `json:"skadn_supported"`
+	MRAIDSupported bool     `json:"mraid_supported"`
+	BidFloor       *float64 `json:"bid_floor"`
```
Comment (umutdemirel) [inline openrtb_ext/imp_tapjoy.go:19]: Check out
```
Imp        TJImp        `json:"imp"`
```
If you want to update

**Review by umutdemirel — COMMENTED** [review]
Similar comments as go-prebid

File: adapters/unicorn/unicorntest/exemplary/banner-app-with-ip.json (hunk 2 above)
Comment (mannychu) [reply adapters/unicorn/unicorntest/exemplary/banner-app-with-ip.json]: Added omit empty to bid_floor locations. Removed bid_floor:null test field.

File: adapters/pubmatic/pubmatic.go [hunk 4]
```diff
@@ -383,14 +373,9 @@ func (a *PubmaticAdapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *ad
...
-	if request.Device != nil && request.Device.Geo != nil && strings.ToLower(request.Device.Geo.Country) == "usa" {
-		if impData.pubmatic.Reward == 1 && request.Imp[0].BidFloor < 10.0 {
-			request.Imp[0].BidFloor = 10.0
-		}
-		if impData.pubmatic.Reward != 1 && request.Imp[0].BidFloor < 8.0 {
-			request.Imp[0].BidFloor = 8.0
-		}
+	// Overwrite BidFloor if present
+	if impData.pubmatic.BidFloor != nil {
+		request.Imp[0].BidFloor = *impData.pubmatic.BidFloor
```
Comment (umutdemirel) [inline adapters/pubmatic/pubmatic.go:378]: For a smooth transition of a feature, we should keep both logic in place initially, and after deploying everything related to this change, we should visit here and deprecate old logic. This is the best approach.

For this one, it's not much an important to pass a different value for a short amount of time, so we can ignore. But for critical features like cache migrations or so, please keep that in mind.

File: adapters/pubmatic/pubmatictest/exemplary/dsp-based-bid-floor-banner.json [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                      {
+                         "key": "preference",
+                         "value": ["sports", "movies"]
+                      }
+                    ],
+                    "wrapper": {
+                        "version": 1,
+                        "profile": 5123
+                    },
+                    "mraid_supported": true
```
Comment (umutdemirel) [inline adapters/pubmatic/pubmatictest/exemplary/dsp-based-bid-floor-banner.json:30]: can't see floor override here, what are you testing? Or did you forgot to add an override floor and update expected request down below?

**Review by umutdemirel — APPROVED** [review]
1 test comment, 1 good practice comment, other than that LGTM!

File: adapters/pubmatic/pubmatic.go (hunk 4 above)
Comment (mannychu) [reply adapters/pubmatic/pubmatic.go:378]: Got it! Thanks for the info, I'll keep this in mind in the future

File: adapters/pubmatic/pubmatictest/exemplary/dsp-based-bid-floor-banner.json (hunk 5 above)
Comment (mannychu) [reply adapters/pubmatic/pubmatictest/exemplary/dsp-based-bid-floor-banner.json]: I used the wrong override in the wrong place, I added `bid_floor` in the imp.ext.bidder in the request and added the corresponding value in the response imp
